// angular-playground/dist/playground/esm2015/lib/core/shared/url.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Location } from '@angular/common';
import { SandboxLoader } from './sandbox-loader';
let UrlService = class UrlService {
    constructor(location) {
        this.location = location;
        this._embed = null;
        this._select = null;
        this.sandboxMenuItems = SandboxLoader.getSandboxMenuItems();
        const urlPath = location.path();
        this._embed = /[?|&]embed=1/.exec(urlPath) !== null;
        this._select = this.parse('scenario', this.sandboxMenuItems, urlPath);
        if (this._select) {
            if (this._select.sandboxKey === null && this._select.scenarioKey === null) {
                this.location.replaceState('');
                return;
            }
        }
    }
    get embed() {
        return this._embed;
    }
    get select() {
        return this._select;
    }
    setSelected(sandboxKey, scenarioKey) {
        if (sandboxKey === null && scenarioKey === null) {
            this.location.replaceState('');
            return;
        }
        const sandBoxMenuItem = this.sandboxMenuItems
            .find(sandboxMenuItem => sandboxMenuItem.key.toLowerCase() === sandboxKey.toLowerCase());
        const scenarioMenuItem = sandBoxMenuItem.scenarioMenuItems.find(mi => mi.key === scenarioKey);
        const key = sandBoxMenuItem.uniqueId ? sandBoxMenuItem.uniqueId : sandboxKey;
        this.location.replaceState(`?scenario=${encodeURIComponent(key)}/${encodeURIComponent(scenarioMenuItem.description)}`);
    }
    parse(key, sandboxMenuItems, urlPath) {
        const match = new RegExp('[?|&]' + key + '=([^&#]*)').exec(urlPath);
        if (match !== null) {
            const value = decodeURIComponent(match[1]);
            const lastSlash = value.lastIndexOf('/');
            const sbKey = value.substr(0, lastSlash);
            let sandboxKey = sbKey;
            const sandboxMenuItem = sandboxMenuItems
                .find(smi => smi.key.toLowerCase() === sandboxKey.toLowerCase()
                || (smi.uniqueId && smi.uniqueId === sandboxKey));
            sandboxKey = sandboxMenuItem.key;
            if (!sandboxMenuItem) {
                return { sandboxKey: null, scenarioKey: null };
            }
            const scenarioDesc = value.substr(lastSlash + 1, value.length).toLowerCase();
            const scenarioKey = sandboxMenuItem.scenarioMenuItems
                .findIndex(scenarioMenuItem => scenarioMenuItem.description.toLowerCase() === scenarioDesc) + 1;
            if (scenarioKey <= 0) {
                return { sandboxKey: null, scenarioKey: null };
            }
            return { sandboxKey, scenarioKey };
        }
    }
};
UrlService.ctorParameters = () => [
    { type: Location }
];
UrlService = __decorate([
    Injectable()
], UrlService);
export { UrlService };
