// angular-playground/dist/playground/esm2015/lib/core/playground.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { PlaygroundCommonModule } from './playground-common.module';
import { MIDDLEWARE } from '../lib/middlewares';
import { initializePlayground } from '../lib/initialize-playground';
import { BehaviorSubject } from 'rxjs';
const _middleware = new BehaviorSubject({
    selector: null,
    overlay: false,
    modules: [],
});
const middleware = _middleware.asObservable();
const ɵ0 = middleware;
let PlaygroundModule = class PlaygroundModule {
    static configure(configuration) {
        initializePlayground(configuration.selector);
        _middleware.next(Object.assign(Object.assign({}, _middleware.value), configuration));
        return this;
    }
};
PlaygroundModule = __decorate([
    NgModule({
        imports: [
            BrowserModule,
            PlaygroundCommonModule,
        ],
        providers: [
            { provide: MIDDLEWARE, useValue: ɵ0 },
        ],
        exports: [
            PlaygroundCommonModule
        ]
    })
], PlaygroundModule);
export { PlaygroundModule };
export { ɵ0 };
