// angular-playground/dist/playground/esm5/lib/core/playground.module.js
import { __assign, __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { PlaygroundCommonModule } from './playground-common.module';
import { MIDDLEWARE } from '../lib/middlewares';
import { initializePlayground } from '../lib/initialize-playground';
import { BehaviorSubject } from 'rxjs';
var _middleware = new BehaviorSubject({
    selector: null,
    overlay: false,
    modules: [],
});
var middleware = _middleware.asObservable();
var ɵ0 = middleware;
var PlaygroundModule = /** @class */ (function () {
    function PlaygroundModule() {
    }
    PlaygroundModule.configure = function (configuration) {
        initializePlayground(configuration.selector);
        _middleware.next(__assign(__assign({}, _middleware.value), configuration));
        return this;
    };
    PlaygroundModule = __decorate([
        NgModule({
            imports: [
                BrowserModule,
                PlaygroundCommonModule,
            ],
            providers: [
                { provide: MIDDLEWARE, useValue: ɵ0 },
            ],
            exports: [
                PlaygroundCommonModule
            ]
        })
    ], PlaygroundModule);
    return PlaygroundModule;
}());
export { PlaygroundModule };
export { ɵ0 };
