// angular-playground/dist/playground/esm5/lib/core/drawer/drawer.component.js
import { __decorate } from "tslib";
import { Component, EventEmitter, Output } from '@angular/core';
var DrawerComponent = /** @class */ (function () {
    function DrawerComponent() {
        this.openCommandBarClick = new EventEmitter();
    }
    DrawerComponent.prototype.openCommandBar = function () {
        this.openCommandBarClick.emit();
    };
    __decorate([
        Output()
    ], DrawerComponent.prototype, "openCommandBarClick", void 0);
    DrawerComponent = __decorate([
        Component({
            selector: 'ap-drawer',
            template: "<a\n    (click)=\"openCommandBar()\"\n    class=\"link\">\n    Menu\n</a>\n",
            styles: [".link{-webkit-box-align:center;align-items:center;background:#252526;border-radius:50%;bottom:20px;box-shadow:0 0 8px 2px rgba(0,0,0,.5);color:rgba(255,255,255,.5);cursor:pointer;display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column;font-family:Arial,sans-serif;font-size:8px;font-weight:700;height:45px;-webkit-box-pack:center;justify-content:center;position:fixed;right:20px;text-align:center;text-transform:uppercase;-webkit-transition:opacity .5s,-webkit-transform .1s;transition:transform .1s,opacity .5s,-webkit-transform .1s;width:45px;z-index:10000}.link::before{background:-webkit-gradient(linear,left top,left bottom,from(rgba(255,255,255,.5)),color-stop(20%,rgba(255,255,255,.5)),color-stop(20%,transparent),color-stop(40%,transparent),color-stop(40%,rgba(255,255,255,.5)),color-stop(60%,rgba(255,255,255,.5)),color-stop(60%,transparent),color-stop(80%,transparent),color-stop(80%,rgba(255,255,255,.5)),to(rgba(255,255,255,.5)));background:linear-gradient(to bottom,rgba(255,255,255,.5),rgba(255,255,255,.5) 20%,transparent 20%,transparent 40%,rgba(255,255,255,.5) 40%,rgba(255,255,255,.5) 60%,transparent 60%,transparent 80%,rgba(255,255,255,.5) 80%,rgba(255,255,255,.5) 100%);content:'';height:10px;margin-bottom:4px;width:16px}.link:hover{-webkit-transform:scale(1.15);transform:scale(1.15)}:host(.content__menu--hide) .link{opacity:0}"]
        })
    ], DrawerComponent);
    return DrawerComponent;
}());
export { DrawerComponent };
